src/libtiled/mapwriter.cpp

```diff
@@ -618,74 +652,100 @@ void MapWriterPrivate::writeObjectGroup(QXmlStreamWriter &w,
     w.writeEndElement();
 }
 
+static bool shouldWrite(bool holdsInfo, bool isTemplateInstance, bool changed) {
+    return (!isTemplateInstance && holdsInfo) ||
+           (isTemplateInstance && changed);
+}
+
 void MapWriterPrivate::writeObject(QXmlStreamWriter &w,
                                    const MapObject &mapObject)
 {
     w.writeStartElement(QLatin1String("object"));
     const int id = mapObject.id();
     const QString &name = mapObject.name();
     const QString &type = mapObject.type();
-    if (id != 0)
+
+    const QPointF pos = mapObject.position();
+
+    TemplateRef templateRef = mapObject.templateRef();
+    const int templateId = templateRef.templateId;
+    const auto group = templateRef.templateGroup;
```

Comment:
You're not frequently referring to these variables, so please just refer to them using `templateRef.templateId` and `templateRef.templateGroup` instead of having these locals.